src/finalfusion/embeddings.py

```diff
@@ -0,0 +1,563 @@
+"""
+Finalfusion Embeddings
+"""
+from os import PathLike
+from typing import Optional, Tuple, List, Union
+
+import numpy as np
+
+from finalfusion.io import Chunk, Header, _read_chunk_header, ChunkIdentifier
+from finalfusion.metadata import Metadata
+from finalfusion.norms import Norms
+from finalfusion.storage import Storage, NdArray
+from finalfusion.vocab import Vocab, SimpleVocab
+
+
+class Embeddings:  # pylint: disable=too-many-instance-attributes
+    """
+    Embeddings class.
+
+    Typically consists of a :class:`~finalfusion.storage.storage.Storage` and
+    :class:`~finalfusion.vocab.vocab.Vocab`. Other possible chunks are
+    :class:`~finalfusion.norms.Norms` corresponding to the embeddings of the in-vocab tokens and
+    :class:`~finalfusion.metadata.Metadata`.
+
+    If a vocabulary, storage are provided, embeddings can be retrieved through three methods:
+
+    1. :meth:`Embeddings.embedding` allows to provide a default value and returns
+       this value if no embedding could be found.
+    2. :meth:`Embeddings.__getitem__` retrieves an embedding for the query but
+       raises an exception if it cannot retrieve an embedding.
+    3. :meth:`Embeddings.embedding_with_norm` requires a :class:`~finalfusion.norms.Norms`
+       chunk and returns an embedding together with the corresponding L2 norm.
+
+    Embeddings wrap any combination of the 4 chunk types:
+
+    1. :class:`~finalfusion.storage.Storage`:
+        * :class:`~finalfusion.storage.ndarray.NdArray`
+    2. :class:`~finalfusion.vocab.Vocab`:
+        * :class:`~finalfusion.vocab.simple_vocab.SimpleVocab`
+    2. :class:`~finalfusion.metadata.Metadata`
+    3. :class:`~finalfusion.norms.Norms`
+
+    Examples
+    --------
+    >>> storage = NdArray(np.float32(np.random.rand(2, 10)))
+    >>> vocab = SimpleVocab(["Some", "words"])
+    >>> metadata = Metadata({"Some": "value", "numerical": 0})
+    >>> norms = Norms(np.float32(np.random.rand(2)))
+    >>> embeddings = Embeddings(storage=storage, vocab=vocab, metadata=metadata, norms=norms)
+    >>> embeddings.vocab.words
+    ['Some', 'words']
+    >>> np.allclose(embeddings["Some"], storage[0])
+    True
+    >>> try:
+    ...     embeddings["oov"]
+    ... except KeyError:
+    ...     True
+    True
+    >>> _, n = embeddings.embedding_with_norm("Some")
+    >>> np.isclose(n, norms[0])
+    True
+    >>> embeddings.metadata
+    {'Some': 'value', 'numerical': 0}
+    """
+    def __init__(self,
+                 storage: Optional[Storage] = None,
```

Comment:
Do you want to permit storage or vocab to be None?

Reply:
I'm a bit torn here. Having the `Embeddings.embedding` method blow up because the `Embeddings` in fact don't have any embeddings is probably unexpected. 

But I'd like to cover scenarios like vocab + metadata, too. Serialization of single chunks is easy with the `Chunk.write` implementations, but tying them together relies on some container. So I'm not sure how to do both, do you have a suggestion here?

Reply:
Having an embedding file with just a vocab + metadata is advanced use (and probably not what we consider a valid finalfusion file). Users who do stuff like that can read/write the chunks themselves.

Reply:
I'm convinced that `Embeddings` should be well-formed embeddings. I pushed a change that reflects that.

I didn't see the edit in time, as of now, the reader can read in arbitrary order but serialization is in the expected chunk order:

Metadata
Vocab
Storage
Norms

I guess I'll also fix the order for the reader then.

Initially, I wanted to rely on the ChunkIDs specified in the Header, but it turns out that most pretrained embeddings didn't write the ChunkID for the Norms chunk into the header. That could be fixed with a single roundtrip with up-to-date finalfusion-utils but that doesn't catch already downloaded files in the wild.